raft topology: reject replace if the node being replaced is not dead

patjed41 opened the pull request:
The replace operation is defined to succeed only if the node being
replaced is dead. We should reject this operation when the failure
detector considers the node being replaced alive.

Apart from adding this change, this PR adds a test case -
`test_replacing_alive_node_fails` - that verifies it. A few testing
framework adjustments were necessary to implement this test and
to avoid flakiness in other tests that use the replace operation after
the change. From now, we need to ensure that all nodes see the
node being replaced as dead before starting the replace. Otherwise,
the check added in this PR could reject the replace.

Additionally, this PR changes the replace procedure in a way that
if the replacing node reuses the IP of the node being replaced, other
nodes can see it as alive only after the topology coordinator accepts
its join request. The replacing node may become alive before the
topology coordinator checks if the node being replaced is dead. If
that happens and the replacing node reuses the IP of the node being
replaced, the topology coordinator cannot know which of these two
nodes is alive and whether it should reject the join request.

Fixes #15863

avikivity commented:
missing: test

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - Build
:white_check_mark: - Sanity Tests
:white_check_mark: - Unit Tests
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4495/
- Duration: 6 hr 35 min
- Builder: monster

kostja commented:
Can we have a test please?

kostja reviewed (changes requested):
Can we have a test please?

patjed41 commented:
I'm currently working on adding a test.

A natural solution is to add a test case to `test_replace.py`. However, there are already many test cases, and all of them unnecessarily start a new cluster, so I tried to merge them into one test case. Then, I discovered #15967.

I also discovered a small bug in `topology_coordinator::validate_joining_node`. The condition in `if (node.rs->state == node_state::replacing)` is incorrect. `node.rs->state` always evaluates to `node_state::none` for both join and replace. The fix is very simple -- just changing the condition. I'll add a commit with it.

Finally, the testing framework doesn't allow replacing a running node. I need to make a few changes, and after that, I'll be able to submit a test.

patjed41 commented:
v2:
- left the two commits from v1 unchanged (commits 6, 7)
- fixed a bug in `topology_coordinator::validate_joining_node` (commit 1)
- delayed advertising the replacing node if it reuses IP of the node being replaced (commit 2); without this change checking if the node being replaced is dead was bugged
- added a test (commit 8)
- made a few adjustments in the testing framework necessary to write a test and to avoid flakiness in other tests that use the replace operation (commits 3, 4, 5)
- updated PR's description
- rebased on the updated master

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - Build
:white_check_mark: - Sanity Tests
:white_check_mark: - Unit Tests
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4700/
- Duration: 6 hr 47 min
- Builder: monster

kbr-scylla commented on test/pylib/manager_client.py:
```diff
@@ -176,7 +176,8 @@ async def server_add(self, replace_cfg: Optional[ReplaceConfig] = None,
                          cmdline: Optional[List[str]] = None,
```
Re commit message
```
The first one is that if we want to add a server while expecting
an error, we have to do it in two steps:
- call server_add with the start parameter set to False,
- call server_stop with the expected_error parameter.
```

I think you wanted `server_start` instead of `server_stop`.

kbr-scylla commented on test/topology/test_replace.py:
```diff
@@ -48,3 +49,17 @@ async def test_replace_reuse_ip_using_host_id(manager: ManagerClient) -> None:
     await wait_for_token_ring_and_group0_consistency(manager, time.time() + 30)
+
+@pytest.mark.asyncio
+async def test_replacing_alive_node_fails(manager: ManagerClient) -> None:
```
I'd suggest to enable this test only in a single mode like `dev` to save some resources, we don't really exercise some critical paths here so no need to run it in `debug`.

patjed41 replied:
Right, this test is also not the fastest. I'll move it to a separate file and make it `dev` only.

kbr-scylla commented:
@gleb-cloudius @piodul please review

piodul commented on test/pylib/manager_client.py:
```diff
@@ -191,6 +192,14 @@ async def server_add(self, replace_cfg: Optional[ReplaceConfig] = None,
             if expected_error:
                 data['expected_error'] = expected_error
+
+            # If we replace a node and do not expect errors, we wait until other nodes
+            # see the node being replaced as dead. The replace operation may be
+            # rejected if the node being replaced is considered alive.
+            if replace_cfg and not expected_error:
```
I don't like that we are making the wait conditional on whether we expect an error or not. It's not obvious that it's always what we want - what if we want to write a different test than `test_replacing_alive_node_fails` later which expects replace to fail, but not due to the node still being seen as alive by others?

Instead, I think we should extend `ReplaceConfig` to allow choosing whether we want to wait or not; alternatively, add a parameter to `add_server` which allows to control it.

patjed41 replied:
Sure, I'll change it. I think I'll take the second approach. I don't think this waiting is logically a part of the replace config. Also, `server_start` and `server_restart` already have a similar `wait_others` parameter.

patjed41 replied:
Done.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -2966,7 +2966,7 @@ future<> storage_service::join_token_ring(sharded<db::system_distributed_keyspac
         app_states.emplace(gms::application_state::STATUS, versioned_value::normal(my_tokens));
     }
-    if (replacing_a_node_with_same_ip || replacing_a_node_with_diff_ip) {
+    if (!_raft_topology_change_enabled && (replacing_a_node_with_same_ip || replacing_a_node_with_diff_ip)) {
```
Because we do not need to put tokens into the `gossiper state if _raft_topology_change_enabled == true` and before the patch we did not because `replacing_a_node_with_same_ip` and `replacing_a_node_with_diff_ip` were false.

patjed41 replied:
I think this change should be explained in the commit message. I'll add it.

patjed41 replied:
Done.

patjed41 commented:
v2:
- extended and fixed commit messages
- made waiting until the node being replaced is dead in `ManagerClient` optional
- moved `test_replacing_alive_node_fails` to a separate file and made it `dev` only

patjed41 commented:
rebased on a new master as my next PR will depend on this one

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - Build
:white_check_mark: - Sanity Tests
:white_check_mark: - Unit Tests
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4754/
- Duration: 2 hr 25 min
- Builder: spider8.cloudius-systems.com

kbr-scylla commented on test/pylib/manager_client.py:
```diff
@@ -178,7 +179,8 @@ async def server_add(self, replace_cfg: Optional[ReplaceConfig] = None,
                          start: bool = True,
-                         expected_error: Optional[str] = None) -> ServerInfo:
+                         expected_error: Optional[str] = None,
+                         wait_replaced_dead: bool = True) -> ServerInfo:
```
This looks like it belongs inside `ReplaceConfig` struct

kbr-scylla commented on test/topology/test_cluster_features.py:
```diff
@@ -145,7 +145,7 @@ async def test_joining_old_node_fails(manager: ManagerClient) -> None:
 
     # Try to replace with a node that doesn't support the feature - should fail
-    await manager.server_stop(servers[0].server_id)
+    await manager.server_stop_gracefully(servers[0].server_id)
```
I don't like this.

We should be testing cluster behavior under abrupt failures (e.g. `kill -9`), we can't rely on graceful shutdowns to happen each time.

Each time I review this PR I have more and more of this feeling that we're doing something wrong...

First of all, using gossiper liveness information is of course unreliable. What happens when node is marked down by gossiper, not because the node is down, but is behind a network partition? What will the topology coordinator do in this case? I guess that it will attempt to do the replace operation and ban the node in the process. What if the node comes back in the middle of replace? Will it disrupt the cluster while replace is ongoing? Will it send/receive data path operations? Maybe the correct way to handle it is to ban it *at the beginning* of replace operation, not at the end as it is now (if I remember correctly)? @gleb-cloudius WDYT?

I think it's understood that we should be using this stuff only for improving user experience, not relying on it for safety (which is impossible).

So fine, we can:
- reject replace if node is considered alive
- and still have a mechanism that protects us (like early node ban) in case we replace an actually alive (or just temporarily dead) node.

Please check if we need to do that second thing, if it's not done already (do the experiment with having a replaced node come back to life in the middle / at the beginning of replace). If not, please create another follow-up issue.

Second, regarding the abrupt stop. We should have tests with abrupt stop. At least one. Of course waiting 20 seconds because of the user-experience gossiper-based protection is not an option. I see the following options:
- this 20 seconds delay is failure detector setting, which can be configured. We could move these replace tests into `topology_custom` and start clusters there with custom configuration, reducing the delay. This should go together with disabling the tests in debug mode, where the low delay combined with latencies may cause failure detector "flickering"
- OR disable the protection using a failure injection, but that probably also requires moving the tests to `topology_custom`, to enable the failure injection when cluster starts...
- do the above but only to some tests (like the replace tests), the non-replace-specific tests here can keep using `server_stop_gracefully`.

Note I don't want to increase the scope of this PR so that I would also leave to a follow-up issue, which is probably low priority, but still it's an issue. Please create one after the PR is merged.

gleb-cloudius replied:
> I don't like this.
> 
> We should be testing cluster behavior under abrupt failures (e.g. `kill -9`), we can't rely on graceful shutdowns to happen each time.
> 
> Each time I review this PR I have more and more of this feeling that we're doing something wrong...
> 
> First of all, using gossiper liveness information is of course unreliable. What happens when node is marked down by gossiper, not because the node is down, but is behind a network partition? What will the topology coordinator do in this case? I guess that it will attempt to do the replace operation and ban the node in the process. What if the node comes back in the middle of replace? Will it disrupt the cluster while replace is ongoing? Will it send/receive data path operations? Maybe the correct way to handle it is to ban it _at the beginning_ of replace operation, not at the end as it is now (if I remember correctly)? @gleb-cloudius WDYT?

Of course. This was always the final plan. We also need to ban all nodes in the ignored_node list. We just do not have a mechanism to do it since the one we have relies on the `left` state (IIRC).

kbr-scylla replied:
If something is in "the final plan", we should have an issue about it, not keep it in one or two people's heads. (It definitely skipped my awareness until now.) I don't think we have one. @patjed41 please do create an issue about this.

gleb-cloudius replied:
> If something is in "the final plan", we should have an issue about it, not keep it in one or two people's heads. (It definitely skipped my awareness until now.) I don't think we have one. @patjed41 please do create an issue about this.

At least ignored_nodes isolation is in the failure recovery doc:

```
def topology_change():
   isolate_nodes(ignored_nodes)
   version++
```

The plan was to reuse the same `isolate_nodes` that we were going to use during a failure while recovering from a previous one, but in the end that was postponed/canceled, so we did not write `isolate_nodes`.

patjed41 replied:
Do we have the same plan for removenode? That is, banning the node being removed and the ignored nodes. If yes, I think we can create one issue for both replace and removenode.

If I remember correctly, we also want to reject removenode if the node being removed is alive. We already do this in `storage_service::raft_removenode`, but I think we should also (or only) do it on the topology coordinator's side since the gossip state might change. If I'm right, should I create an issue for this?

gleb-cloudius replied:
> Do we have the same plan for removenode?

Of course.

> That is, banning the node being removed and the ignored nodes. If yes, I think we can create one issue for both replace and removenode.
> 
> If I remember correctly, we also want to reject removenode if the node being removed is alive. We already do this in `storage_service::raft_removenode`, but I think we should also (or only) do it on the topology coordinator's side since the gossip state might change. If I'm right, should I create an issue for this?

Yes.

patjed41 replied:
> Second, regarding the abrupt stop. We should have tests with abrupt stop. At least one. Of course waiting 20 seconds because of the user-experience gossiper-based protection is not an option. I see the following options:
> - this 20 seconds delay is failure detector setting, which can be configured. We could move these replace tests into `topology_custom` and start clusters there with custom configuration, reducing the delay. This should go together with disabling the tests in debug mode, where the low delay combined with latencies may cause failure detector "flickering"
> - OR disable the protection using a failure injection, but that probably also requires moving the tests to `topology_custom`, to enable the failure injection when cluster starts...
> - do the above but only to some tests (like the replace tests), the non-replace-specific tests here can keep using `server_stop_gracefully`.

I like the dot 1 + dot 3 approach the most. By "the replace tests" you probably mean `test_replace.py` and `test_raft_ignore_nodes.py`. That will leave us with only one `server_stop -> server_stop_gracefully` change but it's fine.

patjed41 replied:
I opened #16108 and #16109

kbr-scylla replied:
> I opened https://github.com/scylladb/scylladb/issues/16108 and https://github.com/scylladb/scylladb/issues/16109

Great, thanks.

patjed41 commented:
v4:
- moved `wait_replaced_dead` to `ReplaceConfig`
- moved `test_replace.py` to `topology_custom`
- made all initial servers in `test_replace.py` and `test_raft_replace_ignore_nodes` use `failure_detector_timeout_in_ms = 2000`
- changed `server_stop_gracefully` calls back to `server_stop` in `test_replace.py` and `test_raft_replace_ignore_nodes`
- made `test_replace.py` skipped in the debug mode

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - Build
:white_check_mark: - Sanity Tests
:white_check_mark: - Unit Tests
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4792/
- Duration: 2 hr 8 min
- Builder: spider5.cloudius-systems.com

kbr-scylla commented:
Waiting for next to stabilize before merged.
